File: CleanJava/app/src/main/java/com/vascome/fogtail/screens/main/CollectionPresenter.java
package com.vascome.fogtail.screens.main;

import android.support.annotation.NonNull;

import com.vascome.fogtail.screens.base.domain.UseCase;
import com.vascome.fogtail.screens.base.domain.UseCaseHandler;
import com.vascome.fogtail.screens.base.presenters.BasePresenter;
import com.vascome.fogtail.screens.main.domain.model.RecAreaItem;
import com.vascome.fogtail.screens.main.domain.usecase.GetItemsTask;
import com.vascome.fogtail.utils.AnalyticsModel;

import java.util.List;

/**
 * Created by vasilypopov on 12/6/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

public class CollectionPresenter extends BasePresenter<CollectionContract.View> implements CollectionContract.Presenter {

    private final GetItemsTask usecase;
    private final AnalyticsModel analyticsModel;
    private final CollectionContract.Router activityRouter;
    private final UseCaseHandler useCaseHandler;

    public CollectionPresenter(@NonNull GetItemsTask usecase,
                               @NonNull AnalyticsModel analyticsModel,
                               @NonNull CollectionContract.Router activityRouter,
                               @NonNull UseCaseHandler useCaseHandler) {
        this.usecase = usecase;
        this.analyticsModel = analyticsModel;
        this.activityRouter = activityRouter;
        this.useCaseHandler = useCaseHandler;
    }

    @Override
    public void reloadItems() {

        {
            // Tip: in Kotlin you can use ? to operate with nullable values.
            final CollectionContract.View view = view();

            if (view != null) {
                view.setLoadingIndicator(true);
            }
        }
        GetItemsTask.RequestValues requestValue = new GetItemsTask.RequestValues();
        useCaseHandler.execute(usecase, requestValue,
                new UseCase.UseCaseCallback<GetItemsTask.ResponseValue>() {
                    @Override
                    public void onSuccess(GetItemsTask.ResponseValue response) {
                        List<RecAreaItem> items = response.getItems();
                        // Tip: in Kotlin you can use ? to operate with nullable values.
                        final CollectionContract.View view = view();

                        if (view != null) {
                            if(items != null && items.size() > 0) {
                                view.showItems(items);
                            }
                            else {
                                view.showError();
                            }
                        }
                    }

                    @Override
                    public void onError(Throwable t) {
                        analyticsModel.sendError("CollectionPresenter.reloadItems failed", t);
                        // Tip: in Kotlin you can use ? to operate with nullable values.
                        final CollectionContract.View view = view();
                        if (view != null) {
                            view.showError();
                        }
                    }
                });
    }

    @Override
    public void openItemDetail(@NonNull RecAreaItem item) {
        activityRouter.openDetailForItem(item);
    }
}

File: MosbyMVPConductorKotlin/app/src/main/java/com/vascome/fogtail/presentation/main/CollectionPresenter.kt
package com.vascome.fogtail.presentation.main

import com.hannesdorfmann.mosby3.mvp.MvpBasePresenter
import com.vascome.fogtail.presentation.main.domain.model.RecAreaItem
import com.vascome.fogtail.presentation.main.domain.usecase.LoadItemsUseCase
import com.vascome.fogtail.utils.AnalyticsModel

import javax.inject.Inject

import io.reactivex.observers.DisposableObserver

/**
 * Created by vasilypopov on 12/6/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

class CollectionPresenter
@Inject constructor(private val usecase: LoadItemsUseCase,
            private val analyticsModel: AnalyticsModel)
    : MvpBasePresenter<CollectionContract.View>(), CollectionContract.Presenter {

    override var isInProgress: Boolean = false

    private fun showViewLoading() {
        ifViewAttached{ view->view.setLoadingIndicator(true) }
    }

    private fun hideViewLoading() {
        ifViewAttached{ view->view.setLoadingIndicator(false) }
        isInProgress = false
    }

    private fun showErrorMessage() {
        ifViewAttached{view->view.showError()}
    }

    private fun showItems(items: List<RecAreaItem>) {
        ifViewAttached{ view->view.showItems(items) }
    }


    override fun reloadItems() {
        showViewLoading()
        isInProgress = true

        usecase.execute(object : DisposableObserver<List<RecAreaItem>>() {
            override fun onNext(items: List<RecAreaItem>) {
                showItems(items)
            }

            override fun onError(e: Throwable) {
                analyticsModel.sendError("error", e)
                hideViewLoading()
                showErrorMessage()
            }

            override fun onComplete() {
                hideViewLoading()
            }

        } , null)
    }

    override fun destroy() {
        usecase.dispose()
        super.destroy()
    }
}
